<launch>

  <arg name="uav_name" default="$(optenv UAV_NAME uav)"/>
  <arg name="uav_mass" default="$(optenv UAV_MASS uav)"/>
  <arg name="world_file" default="$(find mrs_main)/config/world_simulation.yaml"/>

  <!-- Nodelet Manager -->
  <include file="$(find mrs_main)/launch/nodelet_manager.launch" />

  <!-- Odometry -->
  <include file="$(find mrs_odometry)/launch/simulation.launch" />

  <group ns="$(arg uav_name)">

    <!-- MAV Manager -->
    <node pkg="nodelet" type="nodelet" name="mav_manager" args="load mrs_mav_manager/MavManager $(arg uav_name)_nodelet_manager" output="screen">

      <rosparam file="$(find mrs_mav_manager)/config/mav_manager.yaml" />

      <!-- Subscribers -->
      <remap from="~odometry_in" to="odometry/new_odom" />
      <remap from="~mavros_odometry_in" to="mavros_interface/converted_mavros_odom" />
      <remap from="~tracker_status_in" to="control_manager/tracker_status" />

      <!-- Publishers -->
      <remap from="~attitude_cmd_out" to="mavros/setpoint_raw/attitude" />
      <remap from="~profiler" to="profiler" />

      <!-- Services -->
      <remap from="~takeoff_in" to="~takeoff" />
      <remap from="~land_in" to="~land" />
      <remap from="~land_home_in" to="~land_home" />
      <remap from="~takeoff_out" to="control_manager/landoff_tracker/takeoff" />
      <remap from="~land_out" to="control_manager/landoff_tracker/land" />
      <remap from="~switch_tracker_out" to="control_manager/switch_tracker" />
      <remap from="~motors_out" to="control_manager/motors" />
      <remap from="~emergency_goto_out" to="control_manager/emergency_goto" />
      <remap from="~enable_callbacks_out" to="control_manager/enable_callbacks" />

    </node>

    <!-- Control manager -->
    <node pkg="nodelet" type="nodelet" name="control_manager" args="load mrs_mav_manager/ControlManager $(arg uav_name)_nodelet_manager" output="screen">

      <rosparam file="$(find mrs_mav_manager)/config/control_manager.yaml" />

      <!-- Controller's configs -->

      <!-- PID -->
      <rosparam ns="pid_controller" file="$(find mrs_controllers)/config/simulation/iris/pid.yaml" />
      <param name="pid_controller/uav_mass" value="$(arg uav_mass)" />
      <remap from="~pid_controller/profiler" to="profiler" />

      <!-- LSF -->
      <rosparam ns="lsf_controller" file="$(find mrs_controllers)/config/simulation/iris/lsf.yaml" />
      <param name="lsf_controller/uav_mass" value="$(arg uav_mass)" />
      <remap from="~lsf_controller/profiler" to="profiler" />

      <!-- FailSafe -->
      <rosparam ns="failsafe_controller" file="$(find mrs_controllers)/config/simulation/iris/failsafe.yaml" />
      <param name="failsafe_controller/uav_mass" value="$(arg uav_mass)" />
      <remap from="~failsafe_controller/profiler" to="profiler" />

      <!-- Trackers' configs -->

      <!-- LineTracker -->
      <rosparam ns="line_tracker" file="$(find mrs_trackers)/config/simulation/line_tracker.yaml" />
      <remap from="~line_tracker/profiler" to="profiler" />

      <!-- LandoffTracker -->
      <rosparam ns="landoff_tracker" file="$(find mrs_trackers)/config/simulation/landoff_tracker.yaml" />
      <remap from="~landoff_tracker/profiler" to="profiler" />

      <!-- MpcTracker -->
      <rosparam ns="mpc_tracker" file="$(find mrs_trackers)/config/simulation/mpc_tracker.yaml" />
      <rosparam ns="mpc_tracker" file="$(arg world_file)" />
      <param name="mpc_tracker/uav_name" value="$(arg uav_name)" />
      <param name="mpc_tracker/predicted_trajectory_topic" value="control_manager/mpc_tracker/predicted_trajectory" />
      <remap from="~mpc_tracker/profiler" to="profiler" />
      <remap from="~mpc_tracker/rc_in" to="mavros/rc/in" />

      <!-- Subscribers -->
      <remap from="~odometry_in" to="odometry/new_odom" />

      <!-- Subscribers and Services -->
      <remap from="~goto_in" to="~goto" />
      <remap from="~goto_relative_in" to="~goto_relative" />
      <remap from="~goto_altitude_in" to="~goto_altitude" />
      <remap from="~set_yaw_in" to="~set_yaw" />
      <remap from="~set_yaw_relative_in" to="~set_yaw_relative" />

      <!-- Publishers -->
      <remap from="~control_output_out" to="mavros/setpoint_raw/attitude" />
      <remap from="~tracker_status_out" to="~tracker_status" />
      <remap from="~cmd_odom_out" to="~cmd_odom" />
      <remap from="~position_cmd_out" to="~position_cmd" />
      <remap from="~attitude_cmd_out" to="~attitude_cmd" />
      <remap from="~profiler" to="profiler" />

      <!-- Services -->
      <remap from="~switch_tracker_in" to="~switch_tracker" />
      <remap from="~switch_controller_in" to="~switch_controller" />
      <remap from="~hover_in" to="~hover" />
      <remap from="~motors_in" to="~motors" />
      <remap from="~emergency_goto_in" to="~emergency_goto" />
      <remap from="~enable_callbacks_in" to="~enable_callbacks" />

    </node>

  </group>

</launch>
